src/mygrad/tensor_base.py

```diff
@@ -527,18 +528,55 @@ def _op(
         vars_can_share_mem = (
             isinstance(var, (np.ndarray, Tensor)) for var in input_vars
         )
-        tensor_vars = tuple(
-            (
-                cls(var, constant=True)
-                if not isinstance(var, Tensor)
-                else var.null_grad()
-            )._lock_writeability()
-            for var in input_vars
-        )
 
+        f = Op()
+
+        if _track.TRACK_GRAPH:
+
+            tensor_vars = tuple(
+                (
+                    cls(var, constant=True)
+                    if not isinstance(var, Tensor)
+                    else var.null_grad()
+                )._lock_writeability()
+                for var in input_vars
+            )
+        else:
+            tensor_vars = tuple(
+                cls(var, constant=True) if not isinstance(var, Tensor) else var
+                for var in input_vars
+            )
+
+        op_out = f(*tensor_vars, *op_args, **op_kwargs)  # type: np.ndarray
+
+        # determine whether or not op was a view
+        if f.cannot_return_view:
+            base = None
+        else:
+            for can_share_mem, var in zip(vars_can_share_mem, tensor_vars):
+                if can_share_mem and _is_view_of(parent=var, child=op_out):
+                    base = var if var.base is None else var.base
+                    assert not f.cannot_return_view, (
+                        f"{f} is marked as being unable to return a view, "
+                        f"however its output is a view of one of its inputs"
+                    )
+                    break
+            else:
+                base = None
```

Comment:
This loop makes me suspicious.

- You could define `vars_can_share_mem` immediately before use; for clarity and to avoid pointless work `if f.cannot_return_view`
- The assertion is redundant, because we're in the else-condition of a branch on that same test

Reply:
Good call on both fronts